File: instacapture/src/main/java/com/tarek360/instacapture/screenshot/FieldHelper.java
package com.tarek360.instacapture.screenshot;

import android.app.Activity;
import android.os.Build;
import android.view.View;
import android.view.WindowManager;
import com.tarek360.instacapture.exception.ScreenCapturingFailedException;
import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.List;

/**
 * Created by tarek on 5/18/16.
 */
class FieldHelper {

  private final static String FIELD_NAME_WINDOW_MANAGER = "mWindowManager";
  private final static String FIELD_NAME_GLOBAL = "mGlobal";
  private final static String FIELD_NAME_ROOTS = "mRoots";
  private final static String FIELD_NAME_PARAMS = "mParams";
  private final static String FIELD_NAME_VIEW = "mView";

  private FieldHelper() {
  }

  public static List<RootViewInfo> getRootViews(Activity activity) {
    List<RootViewInfo> rootViews = new ArrayList<>();

    Object windowManager;
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
      windowManager = getFieldValue(FIELD_NAME_GLOBAL, activity.getWindowManager());
    } else {
      windowManager = getFieldValue(FIELD_NAME_WINDOW_MANAGER, activity.getWindowManager());
    }

    Object rootObjects = getFieldValue(FIELD_NAME_ROOTS, windowManager);
    Object paramsObject = getFieldValue(FIELD_NAME_PARAMS, windowManager);

    Object[] viewRoots;
    WindowManager.LayoutParams[] params;

    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
      viewRoots = ((List) rootObjects).toArray();
      List<WindowManager.LayoutParams> paramsList = (List<WindowManager.LayoutParams>) paramsObject;
      params = paramsList.toArray(new WindowManager.LayoutParams[paramsList.size()]);
    } else {
      viewRoots = (Object[]) rootObjects;
      params = (WindowManager.LayoutParams[]) paramsObject;
    }

    for (int i = 0; i < viewRoots.length; i++) {
      View view = (View) getFieldValue(FIELD_NAME_VIEW, viewRoots[i]);

      rootViews.add(new RootViewInfo(view, params[i]));
    }

    return rootViews;
  }

  private static Object getFieldValue(String fieldName, Object target) {

    try {
      Field field = target.getClass().getDeclaredField(fieldName);
      field.setAccessible(true);
      return field.get(target);
    } catch (Exception e) {
      throw new ScreenCapturingFailedException(e);
    }
  }
}

File: instacapture/src/main/java/com/tarek360/instacapture/screenshot/FieldHelper.kt
package com.tarek360.instacapture.screenshot

import android.app.Activity
import android.os.Build
import android.view.View
import android.view.WindowManager
import java.util.*

/**
 * Created by tarek on 5/18/16.
 */
internal object FieldHelper {

    private val FIELD_NAME_WINDOW_MANAGER = "mWindowManager"
    private val FIELD_NAME_GLOBAL = "mGlobal"
    private val FIELD_NAME_ROOTS = "mRoots"
    private val FIELD_NAME_PARAMS = "mParams"
    private val FIELD_NAME_VIEW = "mView"
    private val FIELD_NAME_VIEWS = "mViews"

    fun getRootViews(activity: Activity): List<RootViewInfo> {
        val rootViews = ArrayList<RootViewInfo>()

        val windowManager: Any
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
            windowManager = getFieldValue(FIELD_NAME_GLOBAL, activity.windowManager)!!
        } else {
            windowManager = getFieldValue(FIELD_NAME_WINDOW_MANAGER, activity.windowManager)!!
        }

        val rootObjects = getFieldValue(FIELD_NAME_ROOTS, windowManager)
        val paramsObject = getFieldValue(FIELD_NAME_PARAMS, windowManager)

        val viewRoots: Array<Any?>
        val params: Array<WindowManager.LayoutParams?>

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
            viewRoots = (rootObjects as List<*>).toTypedArray()
            val paramsList = paramsObject as List<WindowManager.LayoutParams>
            params = paramsList.toTypedArray()
        } else {
            viewRoots = rootObjects as Array<Any?>
            params = paramsObject as Array<WindowManager.LayoutParams?>
        }
        for (i in viewRoots.indices) {
            val view = getFieldValue(FIELD_NAME_VIEW, viewRoots[i])
            if (view == null) {
                val views = getFieldValue(FIELD_NAME_VIEWS, windowManager) as ArrayList<View>
                if (views != null && views.size > 0) {
                    views.filter { it.visibility == View.VISIBLE }
                            .mapTo(rootViews) { RootViewInfo(it, params[i]) }
                }
            } else {
                val view = view as View
                if (view.visibility == View.VISIBLE) {
                    rootViews.add(RootViewInfo(view, params[i]))
                }
            }

        }

        return rootViews
    }

    @Throws(NoSuchFieldException::class, IllegalAccessException::class)
    private fun getFieldValue(fieldName: String, target: Any?): Any? {
        try {
            val field = target?.javaClass?.getDeclaredField(fieldName)
            field?.isAccessible = true
            return field?.get(target)
        } catch (e: Exception) {
            e.printStackTrace()
            return null;
        }
    }
}
